### ISSUE TITLE
Auto Exploitation of Sudo Tokens without flag (LinPEAS)

### ISSUE THREAD

**Issue by MuirlandOracle** [carlospolop/privilege-escalation-awesome-scripts-suite#125]
https://github.com/carlospolop/privilege-escalation-awesome-scripts-suite/blob/28fd74ffe5495aab6a3495c872399383578f37c1/linPEAS/linpeas.sh#L1557-L1579

In the referenced lines, the script attempts to exploit a sudo injection vulnerability; however, this auto-exploitation in an enumeration script (at the default level) feels very inappropriate, as no indication is given prior to running the script that it will attempt to do so.

If possible, please consider adding a switch to activate any auto-exploitation features (or put them under the `-a` switch), leaving the indicator of vulnerability as it is (as is the case elsewhere in the script). It feels wrong having to worry about enumeration scripts attempting to exploit things automatically -- especially given the README states explicitly that Linpeas is designed to "search for privilege escalation paths", and will not write anything to the disk (which this auto-exploit does) or attempt to bruteforce in the default mode.

*As a side note, this auto-exploitation apparently also makes the LinPEAS script very dangerous to use in the OSCP examination due to this feature (see the [attached Twitter thread](https://twitter.com/_superhero1/status/1385206684109447168)), which is obviously a huge blow for anyone attempting to sit the exam due to how widely used this tool is, and the possibility of getting a target vulnerable to the one piece of auto-exploitation code in the script which would result in an immediate fail, as evidenced above.*

**Comment by Omega-Void** [comment]
Since this can't be detected without exploiting it, one possible solution would be something like outputting:
"Fits exploit criteria for sudo token inmpersonation, execute the script with flag -a to attempt to exploit."

**Comment by adithyan-ak** [comment]
This line is the actual problem 
`call system("echo | sudo -S cp /bin/sh /tmp/shrndom >/dev/null 2>&1 && echo | sudo -S chmod +s /tmp/shrndom >/dev/null 2>&1")`
Instead of creating a SUID bash in /tmp, a simple check for root access and displaying the target is vulnerable to `sudo_inject` along with reference link for[ manual exploit ](https://github.com/nongiach/sudo_inject/blob/master/exploit_v2.sh)script would fix this problem.

**Comment by MuirlandOracle** [comment]
I left the issue open ended to leave it up to the tool maintainer exactly what they wanted to do with it, but yes, wrapping that line inside an if/else to check whether the user has asked for exploitation then just returning information on how to exploit it manually if the user hasn't specifically asked for auto-exploitation would be my preferred resolution. :)

**Comment by carlospolop** [comment]
Hi guys!
Thank you for your concerns.
Instead of adding a new parameter I have modified the code so it doesn't give root access but just clearly indicates that the host is vulnerable (this option was also suggested). You can check the new code: https://github.com/carlospolop/privilege-escalation-awesome-scripts-suite/blob/84e9df0addddd9a9cfabf31e517d5a9266f9eb3a/linPEAS/linpeas.sh#L1572-L1581

I hope this will be enough and Offensive Security will allow the use of linpeas again.

Feel free to let me know what you think

**Comment by ajdumanhug** [comment]
Saying hello to superhero stream!

**Comment by superhero1** [comment]
Awesome @carlospolop ! Thank you so much! Greetings from my stream <3

**Comment by carlospolop** [comment]
hahaha :)

**Comment by adithyan-ak** [comment]
Hi @carlospolop , instead of creating /tmp/shrndom32r2r, can't we check by if [[ $(id -u) -eq 0 ]] and echo sudo token reuse worked?

**Comment by Tib3rius** [comment]
@carlospolop depending on how OffSec define "exploit" this fix may still be "illegal" for the exam. Since your code is still exploiting the vulnerability, even just to perform a simple file creation check, that might be enough for them to consider it "auto-exploit". Granted, the exploit doesn't get you a shell as root, but it's still exploiting a vuln.

**Comment by Omega-Void** [comment]
> Hi guys!
> Thank you for your concerns.
> Instead of adding a new parameter I have modified the code so it doesn't give root access but just clearly indicates that the host is vulnerable (this option was also suggested). You can check the new code:
> 
> https://github.com/carlospolop/privilege-escalation-awesome-scripts-suite/blob/84e9df0addddd9a9cfabf31e517d5a9266f9eb3a/linPEAS/linpeas.sh#L1572-L1581
> 
> I hope this will be enough and Offensive Security will allow the use of linpeas again.
> 
> Feel free to let me know what you think

One possible issue that might arise from something like this is if your scope is limited (for instance, you're allowed to enumerate exploits and access low-privilege accounts but for some reason your scope doesn't include accessing accounts with root privileges, just the fact that the script runs the exploit (to verify it), can land you in trouble as you temporarily get access to something out of scope, and are logged doing so.

I understand the need to be thorough, and it's your prerogative, but should we be running exploits without explicit request by the user?

**Comment by adithyan-ak** [comment]
> > Hi guys!
> > Thank you for your concerns.
> > Instead of adding a new parameter I have modified the code so it doesn't give root access but just clearly indicates that the host is vulnerable (this option was also suggested). You can check the new code:
> > https://github.com/carlospolop/privilege-escalation-awesome-scripts-suite/blob/84e9df0addddd9a9cfabf31e517d5a9266f9eb3a/linPEAS/linpeas.sh#L1572-L1581
> > 
> > I hope this will be enough and Offensive Security will allow the use of linpeas again.
> > Feel free to let me know what you think
> 
> One possible issue that might arise from something like this is if your scope is limited (for instance, you're allowed to enumerate exploits and access low-privilege accounts but for some reason your scope doesn't include accessing accounts with root privileges, just the fact that the script runs the exploit (to verify it), can land you in trouble as you temporarily get access to something out of scope, and are logged doing so.
> 
> I understand the need to be thorough, and it's your prerogative, but should we be running exploits without explicit request by the user?

Agreed. I think checking if ptrace is enabled and displaying the possibility of the exploit is more than enough as Ptrace is not enabled by default in most cases due to kernel hardening in Linux. 

```
ptrace_scope="`cat /proc/sys/kernel/yama/ptrace_scope 2>/dev/null`"
  if [ "$ptrace_scope" ] && [ "$ptrace_scope" -eq 0 ]; then echo "/proc/sys/kernel/yama/ptrace_scope is enabled (0) Possible sudo token reuse"
```

**Comment by catmandx** [comment]
I am taking the oscp exam in a few hours, and the only way to be sure (if you want to use LinPeas) is to delete the sudo token part from the script then informing the proctors (or mention the removal in your report). This is not something we can risk. 

Thank you, Carlos, for taking the time to address this issue, but I think it would be better to completely remove the exploit code to avoid further confusion because OffSec themselves said that you told them you removed the exploit code, but in reality you just adjusted the exploit to not give root access, which is still an automatic exploit.

**Comment by carlospolop** [comment]
Hi @adithyan-ak and @catmandx, it's safe to use linpeas in the exam with the changes performed.
It was confirmed directly by offensive security in this post https://www.offensive-security.com/offsec/understanding-pentest-tools-scripts/?utm_content=163426837&utm_medium=social&utm_source=twitter&hss_channel=tw-134994790
